### ISSUE TITLE
Memory leak problem in LSTM and RNN 

### ISSUE THREAD

**Issue by rasoolfa** [pytorch/pytorch#2198]
Hi @apaszke ,

I guess this problem still exists even on the master.
I followed the https://discuss.pytorch.org/t/tracking-down-a-suspected-memory-leak/1130/42 as well but I don't think so it is completely solved. My Pytorch version is '0.2.0+4a4d884' (got the latest version yesterday)

Take the following as an example:

```
import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np
import torch.autograd as autograd

def Variable(data, *args, **kwargs):
    if torch.cuda.is_available():
        return autograd.Variable(data.cuda(), *args, **kwargs)
    else:
        return autograd.Variable(data, *args, **kwargs)

class LSTM_MEM_LEAK(nn.Module):
    
    def __init__(self):
 
        super(LSTM_MEM_LEAK, self).__init__()
        self.l1 = nn.Linear(256, 300)
        self.lstm =nn.LSTM(300, 400, batch_first = True, num_layers = 1)
        self.l2 = nn.Linear(400, 300)
            
    def forward(self, input):
        
        hidden = (Variable(torch.zeros(1, 16, 400)),
                Variable(torch.zeros(1, 16, 400))) 
        l1 = F.relu(self.l1(input.view(-1, 256)))
        lstm_out, h = self.lstm(l1.view(16, -1, 300), hidden)
        l2  = F.relu(self.l2(lstm_out.contiguous().view(-1, 400)))
        
        return l2    
    
net = LSTM_MEM_LEAK()
net.cuda()
input = Variable(torch.rand(16, 6000, 256))
print(input.requires_grad)
out = net(input)
```
The memory usage before out = net(input) is ~240MB but when this line runs, it goes up to 1708MB. 

Since no gradient will be computed for the input (i.e. requires_grad is False --> so no need to save parameters at the each time steps to calculate the gradient of input), I am not sure why the memory keeps increasing. 
Any comments would be highly appreciated.

**Comment by apaszke** [comment]
Is it really a memory leak or just an unnecessarily high usage? Does it continue to grow if you call it in a loop?

**Comment by rasoolfa** [comment]
Thanks.
I guess it can be considered both. Yes, it does. Look at the following code, it is not specific to LSTM/RNN, it happens  for any recurrent structure:

`
import numpy as np
import torch
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.autograd import Variable

class RNN(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):

        super(RNN, self).__init__()
        self.hidden_size = hidden_size
        self.input_size = input_size
        self.output_size = output_size

        self.i2h = nn.Linear( self.input_size + self.hidden_size, self.hidden_size)
        self.i2o = nn.Linear( self.input_size + self.hidden_size, self.output_size)
        self.softmax = nn.LogSoftmax()

    def forward(self, input, hidden):
        combined = torch.cat((input, hidden), 1)
        hidden = self.i2h(combined)
        output = self.i2o(combined)
        output = self.softmax(output)

        return output, hidden
    def init_hidden(self):
        return Variable(torch.zeros(16, self.hidden_size))

criterion = nn.NLLLoss()
learning_rate = 0.0005
input_size = 500
output_size = 10
n_hidden =  1024
rnn = RNN(input_size, n_hidden,output_size)
rnn.cuda()
hidden =  rnn.init_hidden().cuda()
for i in range(10000):
    data = Variable(torch.rand(16,input_size).cuda())
    out, hidden = rnn(data, hidden)
`

As I mentioned previously, It happens because it keeps "all" parameters at each time step which not are required to calculate the gradient. Take the following model as an example:

`
h1 = WX1 + Uh0
`
the gradient with respect to W, U, X, h0 would be:
'
dW  = dh_1*X1^T
dU  = dh_1*h0^T
**dX  = W^T dh1
dh0 = U^t dh1** 
'
So the autogard is not required to save W, U if no gradient is required to be calculated for X, h0. However, in the current implementation, it saves everything regardless of  "requires_grad" content.

**Comment by lolongcovas** [comment]
For inference you have to:

model.eval() and all the variables should have volatile=True,
Variable(data, volatile =True)

**Comment by rasoolfa** [comment]
But the problem is not the eval time, it is training time. In the eval, the backward()/optimizer.step() will not be called anyway.

**Comment by lolongcovas** [comment]
Yes,  but your code does not seems in training time... Please try it with
volatile. I had similar issues and solved by this.

**Comment by rasoolfa** [comment]
Thanks.
Yeah, I removed the unnecessary parts (e.g. backward(), params updates) to just only emphasize on the problem.

**Comment by Stonesjtu** [comment]
@rasoolfa  Perhaps you should rename this issue like `memory in-efficiency in RNN` rather than `memory leaking`. Although the memory usage is higher than expected, it doesn't(or you don't report yet) keep growing steadily w.r.t iterations.

**Comment by apaszke** [comment]
It should have been fixed in ceb4f84 (commit that was merged day after you pulled from master).

**Comment by rasoolfa** [comment]
Thanks @apaszke.
 Hmm, I have tried with the 0.2.0+04f31aa version [fresh clone a couple of mins ago], but it happens?

**Comment by rasoolfa** [comment]
I guess the following issue is the side effect of this problem:
https://discuss.pytorch.org/t/cudnn-status-bad-param-and-lstm/5604
Anyway, I'll open up a separate ticket for this new one, but posted here in case of more test cases for the current problem.

**Comment by acrosson** [comment]
This appears to be happening when using CPU as well. I'll try building from source like @apaszke suggested. Any idea when this is going to be released?

**Comment by acrosson** [comment]
The memory is still growing for me after building pytorch from source. I tried my own project as well as @rasoolfa's code example, but memory steadily increased until OOM was thrown. @apaszke does [his code](https://github.com/pytorch/pytorch/issues/2198#issuecomment-318123552) work on your end? (i tested without cuda)

**Comment by rasoolfa** [comment]
@acrosson,
[my understanding is that] the issue is still there even if you got the latest build from the source. This was the reason that @soumith re-opened the ticket.

**Comment by Stonesjtu** [comment]
@rasoolfa

I ran your code in my PC.
> GPU: 1070Ti
> CUDA: 8
> CUDNN: 6
> torch version: 0.20.0+bcea678

 ```python
for i in range(10000):
    data = Variable(torch.rand(16,input_size).cuda())
    out, hidden = rnn(data, hidden) # out and hidden will hold the whole history
```
If your code is like this, plz `detach` your results of RNN at a shorter length. Memory is exhausted by storing the backward information of this sequence of length **10000**.

Here's the detailed log for the objects count and memory usage every **1000** iterations.
```log
current memory use: 1.37623977661
                                              types |   # objects |   total size
=================================================== | =========== | ============
                                               dict |        2001 |    547.90 KB
     <class 'torch.autograd.function.ConcatBackward |        1000 |    234.38 KB
      <class 'torch.autograd.function.AddmmBackward |        1000 |    234.38 KB
  <class 'torch.autograd.function.TransposeBackward |        1000 |    234.38 KB
                                               list |        1153 |    115.91 KB
                                         torch.Size |        1000 |     70.31 KB
                                                int |        2038 |     47.77 KB
                                               long |        1000 |     27.34 KB
                                                str |         163 |      9.79 KB
                    <class 'psutil._pslinux.Process |           1 |     80     B
                                              float |           3 |     72     B
                             <class 'psutil.Process |           1 |     64     B
current memory use: 1.38377380371
                                              types |   # objects |   total size
=================================================== | =========== | ============
                                               dict |        2000 |    546.88 KB
     <class 'torch.autograd.function.ConcatBackward |        1000 |    234.38 KB
      <class 'torch.autograd.function.AddmmBackward |        1000 |    234.38 KB
  <class 'torch.autograd.function.TransposeBackward |        1000 |    234.38 KB
                                               list |        1002 |    101.75 KB
                                         torch.Size |        1000 |     70.31 KB
                                                int |        2007 |     47.04 KB
                                               long |        1000 |     27.34 KB
                                                str |           2 |    127     B
current memory use: 1.40835571289
                                              types |   # objects |   total size
=================================================== | =========== | ============
                                               dict |        2000 |    546.88 KB
     <class 'torch.autograd.function.ConcatBackward |        1000 |    234.38 KB
      <class 'torch.autograd.function.AddmmBackward |        1000 |    234.38 KB
  <class 'torch.autograd.function.TransposeBackward |        1000 |    234.38 KB
                                               list |        1000 |    101.56 KB
                                         torch.Size |        1000 |     70.31 KB
                                                int |        2000 |     46.88 KB
                                               long |        1000 |     27.34 KB
current memory use: 1.41101074219
                                              types |   # objects |   total size
=================================================== | =========== | ============
                                               dict |        2000 |    546.88 KB
     <class 'torch.autograd.function.ConcatBackward |        1000 |    234.38 KB
      <class 'torch.autograd.function.AddmmBackward |        1000 |    234.38 KB
  <class 'torch.autograd.function.TransposeBackward |        1000 |    234.38 KB
                                               list |        1000 |    101.56 KB
                                         torch.Size |        1000 |     70.31 KB
                                                int |        2000 |     46.88 KB
                                               long |        1000 |     27.34 KB
current memory use: 1.43473052979
                                              types |   # objects |   total size
=================================================== | =========== | ============
                                               dict |        2000 |    546.88 KB
     <class 'torch.autograd.function.ConcatBackward |        1000 |    234.38 KB
      <class 'torch.autograd.function.AddmmBackward |        1000 |    234.38 KB
  <class 'torch.autograd.function.TransposeBackward |        1000 |    234.38 KB
                                               list |        1000 |    101.56 KB
                                         torch.Size |        1000 |     70.31 KB
                                                int |        2000 |     46.88 KB
                                               long |        1000 |     27.34 KB
current memory use: 1.43745422363
                                              types |   # objects |   total size
=================================================== | =========== | ============
                                               dict |        2000 |    546.88 KB
     <class 'torch.autograd.function.ConcatBackward |        1000 |    234.38 KB
      <class 'torch.autograd.function.AddmmBackward |        1000 |    234.38 KB
  <class 'torch.autograd.function.TransposeBackward |        1000 |    234.38 KB
                                               list |        1000 |    101.56 KB
                                         torch.Size |        1000 |     70.31 KB
                                                int |        2000 |     46.88 KB
                                               long |        1000 |     27.34 KB
current memory use: 1.43940734863
                                              types |   # objects |   total size
=================================================== | =========== | ============
                                               dict |        2000 |    546.88 KB
     <class 'torch.autograd.function.ConcatBackward |        1000 |    234.38 KB
      <class 'torch.autograd.function.AddmmBackward |        1000 |    234.38 KB
  <class 'torch.autograd.function.TransposeBackward |        1000 |    234.38 KB
                                               list |        1000 |    101.56 KB
                                         torch.Size |        1000 |     70.31 KB
                                                int |        2000 |     46.88 KB
                                               long |        1000 |     27.34 KB
current memory use: 1.44160461426
                                              types |   # objects |   total size
=================================================== | =========== | ============
                                               dict |        2000 |    546.88 KB
     <class 'torch.autograd.function.ConcatBackward |        1000 |    234.38 KB
      <class 'torch.autograd.function.AddmmBackward |        1000 |    234.38 KB
  <class 'torch.autograd.function.TransposeBackward |        1000 |    234.38 KB
                                               list |        1000 |    101.56 KB
                                         torch.Size |        1000 |     70.31 KB
                                                int |        2000 |     46.88 KB
                                               long |        1000 |     27.34 KB
current memory use: 1.44356155396
                                              types |   # objects |   total size
=================================================== | =========== | ============
                                               dict |        2000 |    546.88 KB
     <class 'torch.autograd.function.ConcatBackward |        1000 |    234.38 KB
      <class 'torch.autograd.function.AddmmBackward |        1000 |    234.38 KB
  <class 'torch.autograd.function.TransposeBackward |        1000 |    234.38 KB
                                               list |        1000 |    101.56 KB
                                         torch.Size |        1000 |     70.31 KB
                                                int |        2000 |     46.88 KB
                                               long |        1000 |     27.34 KB
current memory use: 1.44561767578
```

**Comment by apaszke** [comment]
As @Stonesjtu noticed, `hidden` will prevent any part of the graph from being freed, so it will be saving state for backward over the whole sequence. Adding a periodical `.detach_()` will give you BPTT and won't fill up the whole memory

**Comment by rasoolfa** [comment]
Thanks, @apaszke and @Stonesjtu . 
You mean hidden.detach()/h._detach() in above codes?

**Comment by Stonesjtu** [comment]
You may refer to this example to get precise usage of `detach` in a long sequence:
https://github.com/pytorch/examples/blob/master/word_language_model/main.py
line132 `hidden = repackage_hidden(hidden)` does the same thing as `detach`.

**Comment by rasoolfa** [comment]
@Stonesjtu  tnx

**Comment by wise-east** [comment]
@Stonesjtu Hi may I know how you visualized the following in your reply?

I would really like to be able to use something like it to monitor my gpu usage in terms of what objects were taking up my memory. Thanks in advance. 


`current memory use: 1.37623977661
                                              types |   # objects |   total size
=================================================== | =========== | ============
                                               dict |        2001 |    547.90 KB
     <class 'torch.autograd.function.ConcatBackward |        1000 |    234.38 KB
      <class 'torch.autograd.function.AddmmBackward |        1000 |    234.38 KB
  <class 'torch.autograd.function.TransposeBackward |        1000 |    234.38 KB
                                               list |        1153 |    115.91 KB
                                         torch.Size |        1000 |     70.31 KB
                                                int |        2038 |     47.77 KB
                                               long |        1000 |     27.34 KB
                                                str |         163 |      9.79 KB
                    <class 'psutil._pslinux.Process |           1 |     80     B
                                              float |           3 |     72     B
                             <class 'psutil.Process |           1 |     64     B`

**Comment by Stonesjtu** [comment]
@wise-east  I'm sorry I just forget that. Actually the printed information does not help when debugging OOM problem.

But to profile the GPU memory usage I would recommend this gist, i wrote it speically for pytorch (https://gist.github.com/Stonesjtu/368ddf5d9eb56669269ecdf9b0d21cbe).

Because Pytorch uses cached memory allocation, you may also want to check out this. (https://pytorch.org/docs/master/notes/cuda.html#memory-management)

**Comment by Stonesjtu** [comment]
@wise-east It's pympler,  check this out https://stackoverflow.com/a/29923561/6759409  !

**Comment by zhixinma** [comment]
maybe you can check your data type of the embedding, it should be float32 but not float64.
This error can arise by a customed embedding matrix.
